File: cses/1086.py

ones = [0] + [(i+1)*10**i for i in range(18)]

def g(d, w):  # solves d*10**w > 0
    if d == 1:
        return ones[w]+1
    return 10**w + d*ones[w]

def f(h):
    s = str(h)
    l = 0
    res = 0
    for i, c in enumerate(s):
        d = int(c)
        if d == 0: continue
        w = len(s)-i-1
        res += g(d, w)
        res += l*d*10**w if l else 0
        if d == 1: l += 1
    return res

n = int(input())

l = 0
r = 10**18
while l < r:
    h = (l+r+1)//2
    if f(h) > n: r = h-1
    else: l = h
print(l)

File: cses/1086.cpp
#include <bits/stdc++.h>
using namespace std;

using ll = long long;

vector<ll> ones, tens;

ll g(int d, int w) {
    if (d == 1) return ones[w]+1;
    return tens[w] + d*ones[w];
}

ll f(ll h) {
    string s = to_string(h);
    int l = 0;
    ll res = 0;
    for (int i = 0; i < s.size(); ++i) {
        int d = s[i]-'0';
        if (d == 0) continue;
        int w = s.size()-i-1;
        res += g(d, w);
        res += l*d*tens[w];
        l += d == 1;
    }
    return res;
}

int main() {
    tens.push_back(1);
    ones.push_back(0);
    for (int i = 0; i < 18; ++i) {
        tens.push_back(tens.back()*10);
        ones.push_back((i+1)*tens[i]);
    }
    ll n; cin >> n;
    ll l = 0, r = 9e17;
    while (l < r) {
        ll h = (l+r+1)/2;
        if (f(h) > n) r = h-1;
        else l = h;
    }
    cout << l << endl;
}